### PR TITLE
UnpartitionedRVD 1

### PR DESCRIPTION
**Pull request by patrick-schultz**
This PR prepares the way to remove `UnpartitionedRVD` and `RVD`, but stops before doing so to keep things simple. The changes are probably easiest to understand one commit at a time, because it actually worked out quite cleanly. I'll describe each commit, to make review easier:
* **remove `TableValue.enforceOrderingRVD`** - I could have removed `enforceOrderingRVD` earlier, as soon as tables were always sorted by key. This also loosens the `TableValue` invariant: Before, we required that if `key=None` then the `rvd` is an `UnpartitionedRVD`, but it was okay to have `key=Some(..)` and an `UnpartitionedRVD`. Then, once all tables were sorted by key, I strengthened it to `key=None` if and only if `rvd` is an `UnpartitionedRVD`. Now, I'm weakening it in the other direction, to allow `key=None` with `rvd` an `OrderedRVD`, possibly with some non-empty key. This is consistent with the rule that the `rvd` must always have a stronger/longer key than the `TableType`.
* **small tweaks** - Now I start working through the `TableIR` nodes, rewriting them to remove explicit uses of `UnpartitionedRVD`. The general plan is to sandwich the rvd logic between `toOrderedRVD` and `toOldStyleRVD`. The first takes an `UnpartitionedRVD` to an `OrderedRVD` with empty key (and leaves `OrderedRVD`s alone), and the second takes an `OrderedRVD` to an `UnpartitionedRVD` if its key was empty, and leaves it alone otherwise. Once they're all rewritten this way, I redefine `toOldStyleRVD` to always return `OrderedRVD`, and `UnpartitionedRVD` is no longer used.
* **remove `TableUnkey`** - With `UnpartitionedRVD` going away, `TableUnkey` is no longer necessary, it's equivalent to keying by an empty key.
* **small tweaks** - these next two rewrite more `TableIR` nodes
* **Merge master** - the big one
* **tweak MatrixColsTable** - 1) Optimize `coerce` by checking if the requested key is empty, avoiding a scan in that case. 2) Optimize `sortedColsValue` by checking if the column key is empty, avoiding the sort in that case. 3) Simplify `colsRVD`, removing the case on the type of the `RVD`, just calling `coerce` and letting the previous optimizations avoid unnecessary work.
* **`distinctByKey` fix** - While looking over `TableIR` implementations, I noticed a bug in `distinctByKey`: you need to be sure no key is split across multiple partitions. To be sure the empty key edge case still works, I added a test to check that `strictify` on an empty-key partitioner will always collapse everything to one partition.
* **Flipped switch** - redifines `toOldStyleRVD` to just return the `OrderedRVD` unchanged, and asserts that `TableValue.rvd` is always an `OrderedRVD`.
* **rest of the `TableIR` tweaks** - added a factory method `OrderedRVD.unkeyed` to replace `UnpartitionedRVD.apply`.
* the rest are simple tidying up

### CODE REVIEW

File: hail/src/main/scala/is/hail/expr/ir/TableIR.scala [hunk 1]
```diff
@@ -65,19 +65,21 @@ case class TableRead(path: String, spec: TableSpec, typ: TableType, dropRows: Bo
...
     val globals = spec.globalsComponent.readLocal(hc, path, typ.globalType)(0)
-    TableValue(typ,
+    val rvd = if (dropRows)
+      OrderedRVD.empty(hc.sc, typ.rvdType)
+    else {
+      val rvd = spec.rowsComponent.read(hc, path, typ.rowType)
+      (rvd, typ.key) match {
+        case (ordered: OrderedRVD, Some(k))
+          if ordered.typ.key.startsWith(k) => ordered
+        case (_, Some(k)) => rvd.toOrderedRVD.changeKey(k)
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/expr/ir/TableIR.scala]: trying to understand what this is doing --

Since the left is `_`, the rvd is an UnpartitionedRVD, so `toOrderedRVD` is a no-op cast, right? Then we coerce?

Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/TableIR.scala]: we should also warn here -- this will help users know when they need to rewrite things.

File: hail/src/main/scala/is/hail/expr/ir/TableValue.scala [hunk 2]
```diff
@@ -15,7 +15,9 @@ import org.json4s.jackson.JsonMethods
 
 case class TableValue(typ: TableType, globals: BroadcastRow, rvd: RVD) {
   require(typ.rowType == rvd.rowType)
-  require(typ.key.isDefined == rvd.isInstanceOf[OrderedRVD])
+//  require(typ.key.isDefined == rvd.isInstanceOf[OrderedRVD])
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/expr/ir/TableValue.scala]: delete

File: hail/src/main/scala/is/hail/rvd/OrderedRVD.scala [hunk 3]
```diff
@@ -684,14 +688,11 @@ class OrderedRVD(
...
-  override def toUnpartitionedRVD: RVD =
-    UnpartitionedRVD(typ.rowType, crdd)
-
-  def toOldStyleRVD: RVD =
-    if (typ.key.isEmpty)
-      toUnpartitionedRVD
-    else
-      this
+  def toOldStyleRVD: RVD = this
+//    if (typ.key.isEmpty)
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/rvd/OrderedRVD.scala]: delete

File: hail/src/main/scala/is/hail/rvd/UnpartitionedRVD.scala [hunk 4]
```diff
@@ -14,20 +14,17 @@ object UnpartitionedRVD {
...
   self =>
 
-  override def toOrderedRVD: OrderedRVD = {
-    new OrderedRVD(
-      typ = OrderedRVDType(FastIndexedSeq.empty, rowType),
-      partitioner = OrderedRVDPartitioner.unkeyed(crdd.getNumPartitions),
-      crdd = crdd
-    )
-  }
+  assert(false)
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/rvd/UnpartitionedRVD.scala]: this shouldn't be here, right? Can still get unpartitioned RVDs from legacy files, right?

File: hail/src/main/scala/is/hail/expr/ir/TableIR.scala (hunk 1 above)
Comment (patrick-schultz) [reply hail/src/main/scala/is/hail/expr/ir/TableIR.scala]: Either the rvd is an UnpartitionedRVD, or it's an OrderedRVD with a key that is not compatible with `k`. The `toOrderedRVD` is always a no-op, it just converts between `UnpartitionedRVD` and the corresponding `OrderedRVD` with empty key. The point is just to let met write all the actual logic as it will be once UnpartitionedRVD is gone. Actually, I shouldn't be doing the match on the type of `rvd`, I'll clean this up.

File: hail/src/main/scala/is/hail/rvd/UnpartitionedRVD.scala (hunk 4 above)
Comment (patrick-schultz) [reply hail/src/main/scala/is/hail/rvd/UnpartitionedRVD.scala]: This PR is to allow me to delete UnpartitionedRVD completely, which I just did in the followup branch. So I wanted to be sure I got rid of all uses.

We should leave in `UnpartitionedRVDSpec`, but read it in as an `OrderedRVD` with empty key. This should have identical semantics.
